packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -102,6 +102,121 @@ const double _inputFormLandscapeHeight = 108.0;
 /// [DatePickerMode.day] mode. It defaults to [DatePickerMode.day], and
 /// must be non-null.
 ///
+/// ### State Restoration
+///
+/// Using this method will not enable state restoration for the date picker.
+/// In order to enable state restoration for a date picker, use
+/// [Navigator.restorablePush] or [Navigator.restorablePushNamed] with
+/// [DatePickerDialog].
+///
+/// For more information about state restoration, see [RestorationManager].
+///
+/// {@tool sample --template=freeform}
+///
+/// This sample demonstrates how to create a restorable Material date picker.
+/// This is accomplished by enabling state restoration by specifying
+/// [MaterialApp.restorationScopeId] and using [Navigator.restorablePush] to
+/// push [DatePickerDialog] when the button is tapped.
+///
+/// {@macro flutter.widgets.RestorationManager}
+///
+/// ```dart imports
+/// import 'package:flutter/material.dart';
+/// ```
+///
+/// ```dart
+/// void main() {
+///   runApp(const MyApp());
+/// }
+///
+/// class MyApp extends StatelessWidget {
+///   const MyApp({Key? key}) : super(key: key);
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return const MaterialApp(
+///       restorationScopeId: 'app',
+///       title: 'Restorable Date Picker Demo',
+///       home: MyHomePage(),
+///     );
+///   }
+/// }
+///
+/// class MyHomePage extends StatefulWidget {
+///   const MyHomePage({Key? key}) : super(key: key);
+///
+///   @override
+///   _MyHomePageState createState() => _MyHomePageState();
+/// }
+///
+/// class _MyHomePageState extends State<MyHomePage> with RestorationMixin {
+///   @override
+///   String get restorationId => 'scaffold_state';
+///
+///   final RestorableDateTime _selectedDate = RestorableDateTime(DateTime.now());
+///   late final RestorableRouteFuture<DateTime> _restorableDatePickerRouteFuture = RestorableRouteFuture<DateTime>(
```

Comment:
I wish there was less boilerplate involved in bringing up a restorable datepicker, but not sure how to do that...

Reply:
I am not that familiar with the state restoration framework, but I agree this seems pretty heavy for someone that just wants the date picker to maintain its state.  I would think most apps just want to call showDatePicker and not have to deal with state restoration.

Given that the showDatePicker function controls the pushing of the dialog route (or at least showDialog() does), is there no way we can handle this internally?

Also, not sure about exposing the `DatePickerDialog` as public. Not opposed to it, but need to think about it as it means more API to support.

Reply:
Yeah, I don't like how clunky this is either and that we have to expose `DatePickerDialog`. We've already done something similar for showDialog to expose state restoration for the other dialogs.

I'm not sure how we can handle this locally either. There are 2 places where I think a lot of the boilerplate comes from:
1) **The need to define the restorable route builder statically.** This prevents dynamic context within the route builder itself, but requires the developer to define the route builder, the arguments that are passed into the route, as the example does.
2) **Defining the restorable route future**. This allows the popped restorable route to properly handle the returned value from popping the route, since the restorable navigator push APIs do not return a Future like the regular ones do.